nifpga/session.py

```diff
@@ -390,8 +413,250 @@ def read(self):
+
+        Returns:
+            decimal.Decimal: Decimal representation of the fixed point number.
+        Returns:
+            (bool, decimal.Decimal): Boolean for overflow, Decimal
+                                     representation of the fixed point number.
+        """
+        buf = self._ctype_type()
+        self._read_func(self._session, self._resource, buf, self._transfer_len)
+        read_array = [elem for elem in buf]
+        fxp_integer = self._combine_array_of_u32_into_one_value(read_array)
+        return self._convert_from_read_value_to_decimal(fxp_integer)
+
+    def _combine_array_of_u32_into_one_value(self, data):
+        """ This method is a helper to convert the array read from hardware
+        and return a single value removing any excessive bits. Because the
+        significant bits are left associated we need to shift the the combined
+        data to the right.
+        For example, if the register had a word length of 54, the 54 MSB would
+        be the fixed point bits. The 10 LSB of the combinedData must be shifted
+        off in order to not mess up further calculations.
+
+        """
+        combinedData = 0
+        for index in range(0, self._transfer_len):
+            combinedData = (combinedData << 32) + data[index]
+        if self._transfer_len > 1:
+            overflow_bit = 1 if self._overflow_enabled else 0
+            total_num_bits = self._word_length + overflow_bit
+            combinedData = combinedData >> (32 * self._transfer_len - total_num_bits)
+        return combinedData
+
+    def _convert_from_read_value_to_decimal(self, data):
+        """ This method converts value from hardware and returns the respective
+        decimal value or a tuple with the overflow status and the decimal
+        value.
+        """
+        overflow = None
+        if self._overflow_enabled:
+            overflow = self._get_overflow_value(data)
+            data = self._remove_overflow_bit(data)
+
+        if self._signed:
+            data = self._integer_twos_comp(data)
+        decimal_value = Decimal(data * self._delta)
+        if self._overflow_enabled:
+            return (overflow, decimal_value)
+        else:
+            return decimal_value
+
+    def _get_overflow_value(self, data):
+        """ Mask out all the data within the word length, leaving the overflow
+        bit. If the result after masking the the word portion of the fixed
+        point is none zero that indicates the data read has overflowed. """
+        mask = 2**(self._word_length)
+        if data & mask > 0:
+            return True
+        return False
+
+    def _remove_overflow_bit(self, data):
+        """ This helper method masks out all bits not inside the word length,
+        ultimately returning a value of data without the overflow bit. """
+        return data & (2**(self._word_length) - 1)
+
+    def _integer_twos_comp(self, data):
+        if (data & (1 << (self._word_length - 1))) != 0:
```

Comment:
What is this conditional checking?

Reply:
Checks the signed bit not sure why updating to do it the same way as I checked overflow